packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts

```diff
@@ -19,33 +17,45 @@ import { Request } from 'express'
 import assert from 'node:assert'
 import { Signer, thresholdCallToSigners } from '../../../common/combine'
 import { BLSCryptographyClient } from '../../../common/crypto-clients/bls-crypto-client'
-import { PromiseHandler } from '../../../common/handlers'
-import { getKeyVersionInfo, requestHasSupportedKeyVersion, sendFailure } from '../../../common/io'
+import { errorResult, ResultHandler } from '../../../common/handlers'
+import { getKeyVersionInfo, requestHasSupportedKeyVersion } from '../../../common/io'
 import { getCombinerVersion, OdisConfig } from '../../../config'
+import { NoQuotaCache } from '../../../utils/no-quota-cache'
+import { AccountService } from '../../services/account-services'
 import { logPnpSignerResponseDiscrepancies } from '../../services/log-responses'
 import { findCombinerQuotaState } from '../../services/threshold-state'
 
-export function createPnpSignHandler(
+export function pnpSign(
   signers: Signer[],
   config: OdisConfig,
-  dekFetcher: DataEncryptionKeyFetcher
-): PromiseHandler<SignMessageRequest> {
+  accountService: AccountService,
+  noQuotaCache: NoQuotaCache
+): ResultHandler<SignMessageRequest> {
   return async (request, response) => {
     const logger = response.locals.logger
-    if (!validateRequest(request)) {
-      sendFailure(WarningMessage.INVALID_INPUT, 400, response)
-      return
+    if (!isValidRequest(request)) {
+      return errorResult(400, WarningMessage.INVALID_INPUT)
     }
 
     if (!requestHasSupportedKeyVersion(request, config, response.locals.logger)) {
-      sendFailure(WarningMessage.INVALID_KEY_VERSION_REQUEST, 400, response)
-      return
+      return errorResult(400, WarningMessage.INVALID_KEY_VERSION_REQUEST)
     }
 
-    if (!(await authenticateUser(request, logger, dekFetcher))) {
-      sendFailure(WarningMessage.UNAUTHENTICATED_USER, 401, response)
-      return
+    const warnings: ErrorType[] = []
+
+    if (!(await authenticateUser(request, logger, accountService.getAccount, warnings))) {
+      return errorResult(401, WarningMessage.UNAUTHENTICATED_USER)
     }
+
+    const account = request.body.account
+    if (noQuotaCache.maximumQuouaReached(account)) {
+      const quota = noQuotaCache.getTotalQuota(account)
+      // can exist a race condition between the hasQuota and getTotalQuota but that's highly improbable
+      if (quota !== undefined) {
+        return errorResult(403, WarningMessage.EXCEEDED_QUOTA)
```

Comment:
shouldn't we pass the quota value to be included in the reponse?

Reply:
I tried to mimic the same response we had from before.
In the case of a 403, we were not sending anything. We can add it, but it will be changing the api, I assumed that we wanted to maintain the same but optimal